src/v/cluster/partition.cc

```diff
@@ -426,6 +429,21 @@ partition::timequery(storage::timequery_config cfg) {
     cfg.max_offset = _raft->get_offset_translator_state()->to_log_offset(
       cfg.max_offset);
     result = co_await _raft->timequery(cfg);
+
+    // It's possible that during the query, our local log was GCed, and the
+    // result may not actually reflect the correct offset for this partition.
+    // If the local log doesn't look like it can serve the query anymore, query
+    // the remote log.
+    if (may_read_from_cloud && _raft->log().start_timestamp() >= cfg.time) {
```

Comment:
The condition was already checked on line 398 and `start_timestamp` can't decrease so if we took the first branch we will take this one as well. Maybe, instead we should check that new segments were added to the manifest after the first `timequery` and if this is the case we should retry the search in the first cloud storage branch? This can be checked quickly by examining last uploaded offset of the manifest.

Reply:
the algorithm in this case will be:
- we need to cache `start_timestamp` of the raft group or `last_uploaded_offset` of the manifest;
- we're checking if the timestamp is below the `start_timestamp` of the raft group and running the `timequery` on cloud data;
- if the `timequery` returned nullopt we need to check if `start_timestamp` of the raft group or the `last_uploaded` have increased then we need to re-run the query.
- if we still don't have the result then we can fallback to `timequery` on the raft group